pycalphad/tests/test_model.py

```diff
@@ -68,4 +69,31 @@ def test_detect_pure_vacancy_phases():
     "Detecting a pure vacancy phase"
     dbf = Database(ZRO2_CUBIC_BCC_TDB)
     with pytest.raises(DofError):
-        Model(dbf,['AL','CU','VA'],'ZRO2_CUBIC')
\ No newline at end of file
+        Model(dbf,['AL','CU','VA'],'ZRO2_CUBIC')
+
+def test_bad_constituents_not_in_model():
+    dbf = Database(TDB_PARAMETER_FILTERS_TEST)
+    mod = Model(dbf, ['A', 'B'], 'ALPHA')
+    assert v.SiteFraction('ALPHA', 0, 'B') not in mod.ast.free_symbols
+
+def test_bad_constituents_do_not_affect_equilibrium():
+    dbf = Database(TDB_PARAMETER_FILTERS_TEST)
+    assert np.isclose(equilibrium(dbf, ['A', 'B'], ['ALPHA'], {v.P: 101325, v.T: 300, v.N: 1, v.X('B'): 0.5}).GM.values.squeeze(), -10.0)
+
+def test_interation_test_method():
+    dbf = Database(TDB_PARAMETER_FILTERS_TEST)
+    interacting_consts = dbf.search(
+        where('parameter'))[4]['constituent_array']
```

Comment:
Is the order of the parameters returned by the search method guaranteed to be in the same order by TinyDB? If not, this test could be a source of flaky, random failures. (This goes for all tests here that assume an order of the parameter list)

Reply:
The default tinydb storage implementation uses dicts, so I would expect order to be preserved as long as the database isn't modified, but this detail should not be relied upon.

Reply:
I think a better way to write this test would be to just check the result (that you get an error when you expect). That way we are free to adjust internal Model APIs without needing to modify tests.

Reply:
@bocklund I have tried many tests and the order of parameters seems to not change. Anyway, I was rewriting the search conditions with the addition of tests to get only the parameters with specified sublattices. With this change, the TDB could be modified with minor chance to influence in tests.
@richardotis until now, the way that I see to make errors be raised due `_array_validity`, `_purity_test` or `_interaction_test` is testing their with phases or comps conditions that would build a phase with no parameter (I am not sure if actually this raises an error, but I really expect that it raises). Do you think that this would be a nice design to it?

Reply:
@igorjrd My preference would be just to calculate the energy at one temperature and composition. If the database returns the wrong parameters, you get the wrong energy. You just have to first make sure that the test fails without your fix.

Reply:
I submitted a new commit that tests the functions using parameters selected by sublattices comps, but I can do some effort to write tests according your suggestion, @richardotis. (I just sent this, because the work was almost done, but it can be reverted).